readthedocs/rtd_tests/tests/test_doc_builder.py

```diff
@@ -0,0 +1,40 @@
+# -*- coding: utf-8 -*-
+from __future__ import (
+    absolute_import, division, print_function, unicode_literals)
+
+from collections import namedtuple
+import tempfile
+
+from django.test import TestCase
+from mock import patch
+
+from readthedocs.doc_builder.backends.sphinx import BaseSphinx
+from readthedocs.projects.exceptions import ProjectConfigurationError
+from readthedocs.projects.models import Project
+
+
+class SphinxBuilderTest(TestCase):
+
+    fixtures = ['test_data']
+
+    def setUp(self):
+        self.project = Project.objects.get(slug='pip')
+        self.version = self.project.versions.first()
+
+        build_env = namedtuple('project', 'version')
+        build_env.project = self.project
+        build_env.version = self.version
+
+        BaseSphinx.type = 'base'
+        BaseSphinx.sphinx_build_dir = tempfile.mkdtemp()
+        self.base_sphinx = BaseSphinx(build_env=build_env, python_env=None)
+
+    @patch('readthedocs.doc_builder.backends.sphinx.BaseSphinx.create_index')
+    @patch('readthedocs.doc_builder.backends.sphinx.BaseSphinx.get_config_params')
+    @patch('readthedocs.doc_builder.backends.sphinx.BaseSphinx.run')
+    @patch('readthedocs.builds.models.Version.get_conf_py_path')
+    def test_create_conf_py(self, get_conf_py_path, run, get_config_params, create_index):
+        create_index.return_value = 'README.rst'
+        get_config_params.return_value = {}
+        get_conf_py_path.side_effect = ProjectConfigurationError
+        self.base_sphinx.append_conf()
```

Comment:
This test is another good use case for `pytest.xfail`:
https://docs.pytest.org/en/latest/skipping.html#xfail-mark-test-functions-as-expected-to-fail

We can set a condition of failing on py3 for something like this.

Also, it's not clear what the test is testing for here? Is there anything to add here that would clarify that the output passed? Perhaps checking the conf file output?

Reply:
Yes, it seems that it lacks of explanation.

This test probably needs the opposite of `assertRaises` since what I'm trying to test here is that even when a `ProjectConfigurationError` is raised when trying to find the `conf.py` (not found), the `conf.py` is auto-generated by RTD and it doesn't fail because of a `TypeError: write() argument must be str, not SafeBytes`

I'm not sure how to write it yet, but I don't think that `xfail` is what we want since it has to pass in both python versions.

Reply:
I've done something like this for this case:

```python
try:
    stuff()
except TypeError:
    pytest.fail('Reasons')
```